net/ipv4/tcp_wave.c

```diff
@@ -414,26 +414,39 @@ static u32 calculate_ack_train_disp(struct wavetcp *ca, const struct rate_sample
 	ca->heuristic_scale = 0;
 
 	/* initialize the value */
-	if (ca->previous_ack_train_disp == 0)
+	if (ca->previous_ack_train_disp == 0) {
 		ca->previous_ack_train_disp = ack_train_disp;
+		}
 	else if (ack_train_disp > ca->previous_ack_train_disp) {
 		/* let's filter the measured value if it is greater than what we
 		 * already have */
 		unsigned long alpha;
 		unsigned long left;
 		unsigned long right;
 
-		alpha = (delta_rtt * AVG_UNIT) / (beta_ms * 1000);
+		alpha = min(AVG_UNIT, ((delta_rtt * AVG_UNIT) / (beta_ms * 1000)));
+			if (alpha ==0){
+			ack_train_disp = ca->previous_ack_train_disp;
+			}
+			else {
 		left = ((AVG_UNIT - alpha) * ca->previous_ack_train_disp) / AVG_UNIT;
 		right = (alpha * ack_train_disp) / AVG_UNIT;
+		
+		DBG("%u [calculate_ack_train_disp] use alpha %lu to filter a received ack_train_disp %u us left %lu right %lu\n", tcp_time_stamp, alpha, 
+ack_train_disp, left, right);
 		ack_train_disp = left + right;
-
-		DBG("%u [calculate_ack_train_disp] use previous ack_train_disp %u us, "
-		    "alpha %lu to filter a received ack_train_disp %u us\n",
-		    tcp_time_stamp, ca->previous_ack_train_disp, alpha, ack_train_disp);
+		}
+				
 	}
-
-	ca->previous_ack_train_disp = ack_train_disp;
+	if (ack_train_disp == 0){
+		ack_train_disp = ca->previous_ack_train_disp;
+	}
+	else {
+		ca->previous_ack_train_disp = ack_train_disp;
+		}
+	
+		DBG("%u [calculate_ack_train_disp] previous_ack_train_disp %u us, final ack_train_disp %u us\n",
+		    tcp_time_stamp, ca->previous_ack_train_disp, ack_train_disp);
 
```

Comment:
This part is redundant because the check over ack_train_disp == 0 is before (and if it's true, we go in heuristic). Not committed